gossip/src/crds_gossip_push.rs

```diff
@@ -164,37 +200,59 @@ impl CrdsGossipPush {
     }
 
     /// process a push message to the network
+    /// Returns origins' pubkeys of upserted values.
     pub(crate) fn process_push_message(
-        &mut self,
+        &self,
         crds: &mut Crds,
         from: &Pubkey,
-        value: CrdsValue,
+        values: Vec<CrdsValue>,
         now: u64,
-    ) -> Result<(), CrdsGossipError> {
-        self.num_total += 1;
-        if !self.wallclock_window(now).contains(&value.wallclock()) {
-            return Err(CrdsGossipError::PushMessageTimeout);
-        }
-        let origin = value.pubkey();
-        self.received_cache
-            .entry(origin)
-            .or_default()
-            .entry(*from)
-            .and_modify(|(_pruned, timestamp)| *timestamp = now)
-            .or_insert((/*pruned:*/ false, now));
-        crds.insert(value, now).map_err(|_| {
-            self.num_old += 1;
-            CrdsGossipError::PushMessageOldVersion
-        })
+    ) -> Vec<Result<Pubkey, CrdsGossipError>> {
+        self.num_total.fetch_add(values.len(), Ordering::Relaxed);
+        let values: Vec<_> = {
+            let wallclock_window = self.wallclock_window(now);
+            let mut received_cache = self.received_cache.lock().unwrap();
+            values
+                .into_iter()
+                .map(|value| {
+                    if !wallclock_window.contains(&value.wallclock()) {
+                        return Err(CrdsGossipError::PushMessageTimeout);
+                    }
+                    let origin = value.pubkey();
+                    let peers = received_cache.entry(origin).or_default();
+                    peers
+                        .entry(*from)
+                        .and_modify(|(_pruned, timestamp)| *timestamp = now)
+                        .or_insert((/*pruned:*/ false, now));
+                    Ok(value)
+                })
+                .collect()
+        };
+        values
+            .into_iter()
+            .map(|value| {
+                let value = value?;
+                let origin = value.pubkey();
+                match crds.insert(value, now) {
+                    Ok(()) => Ok(origin),
+                    Err(_) => {
+                        self.num_old.fetch_add(1, Ordering::Relaxed);
+                        Err(CrdsGossipError::PushMessageOldVersion)
+                    }
+                }
+            })
+            .collect()
```

Comment:
might be missing something, but can this be coalesced together with something like

```
peers
    .entry(*from)
    .and_modify(|(_pruned, timestamp)| *timestamp = now)
    .or_insert((/*pruned:*/ false, now));
let origin = value.pubkey();
match crds.insert(value, now) {
    Ok(()) => Ok(origin),
    Err(_) => {
        self.num_old.fetch_add(1, Ordering::Relaxed);
        Err(CrdsGossipError::PushMessageOldVersion)
    }
}
```

Reply:
I have a follow up change that wraps crds in a RwLock. So here  we need to write-lock crds first, and this is done so to avoid locking crds and received_cache at the same time.